pkg/apis/kubeone/v1beta1/types.go

```diff
@@ -0,0 +1,324 @@
+
+// OpenstackSpec defines the Openstack provider
+type OpenstackSpec struct{}
+
+// PacketSepc defines the Packet cloud provider
+type PacketSpec struct{}
+
+// VsphereSpec defines the vSphere provider
+type VsphereSpec struct{}
+
+// NoneSpec defines a none provider
+type NoneSpec struct{}
+
+// VersionConfig describes the versions of components that are installed on the machines
+type VersionConfig struct {
+	Kubernetes semver.Semver `json:"kubernetes"`
+}
+
+// ClusterNetworkConfig describes the cluster network
+type ClusterNetworkConfig struct {
+	PodSubnet         string `json:"podSubnet"`
+	ServiceSubnet     string `json:"serviceSubnet"`
+	ServiceDomainName string `json:"serviceDomainName"`
+	NodePortRange     string `json:"nodePortRange"`
+	CNI               *CNI   `json:"cni,omitempty"`
+
+	// +optional
+	NetworkID string `json:"networkID,omitempty"`
+}
+
+// CNI config. Only one CNI provider must be used at the single time.
+type CNI struct {
+	Canal    *CanalSpec       `json:"canal"`
+	WeaveNet *WeaveNetSpec    `json:"weaveNet"`
+	External *ExternalCNISpec `json:"external"`
+}
+
+// CanalSpec defines the Canal CNI plugin
+type CanalSpec struct{}
+
+// WeaveNetSpec defines the WeaveNet CNI plugin
+type WeaveNetSpec struct {
+	Encrypted bool `json:"encrypted"`
+}
+
+// ExternalCNISpec defines the external CNI plugin.
+// It's up to the user's responsibility to deploy the external CNI plugin manually or as an addon
+type ExternalCNISpec struct{}
+
+// ProxyConfig configures proxy for the Docker daemon and is used by KubeOne scripts
+type ProxyConfig struct {
+	HTTP    string `json:"http"`
+	HTTPS   string `json:"https"`
+	NoProxy string `json:"noProxy"`
+}
+
+// WorkerConfig describes a set of worker machines
+type WorkerConfig struct {
+	Name     string       `json:"name"`
+	Replicas *int         `json:"replicas"`
+	Config   ProviderSpec `json:"providerSpec"`
+}
+
+// ProviderSpec describes a worker node
+type ProviderSpec struct {
+	CloudProviderSpec   json.RawMessage   `json:"cloudProviderSpec"`
+	Labels              map[string]string `json:"labels"`
+	Taints              []corev1.Taint    `json:"taints,omitempty"`
+	SSHPublicKeys       []string          `json:"sshPublicKeys"`
+	OperatingSystem     string            `json:"operatingSystem"`
+	OperatingSystemSpec json.RawMessage   `json:"operatingSystemSpec"`
+
+	// +optional
+	Network *NetworkConfig `json:"network,omitempty"`
+
+	// +optional
+	OverwriteCloudConfig *string `json:"overwriteCloudConfig,omitempty"`
+}
+
+// DNSConfig contains a machine's DNS configuration
+type DNSConfig struct {
+	Servers []string `json:"servers"`
+}
+
+// NetworkConfig contains a machine's static network configuration
+type NetworkConfig struct {
+	CIDR    string    `json:"cidr"`
+	Gateway string    `json:"gateway"`
+	DNS     DNSConfig `json:"dns"`
+}
+
+// MachineControllerConfig configures kubermatic machine-controller deployment
+type MachineControllerConfig struct {
+	Deploy bool `json:"deploy"`
```

Comment:
I think this field is not needed anymore, and so whole structure can go away.

Reply:
I disagree with this. The operator should be able to choose do they want to use machine-controller or not. We don't depend on machine-controller (you can choose to not deploy workers or use static workers), so it should remain optional.